File: caid/impl/python/run_mapping_vectors.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0

import copy
import json
import os
import sys

from mapping import compare_mapped_actions, mapping_profile_hash

HERE = os.path.dirname(os.path.abspath(__file__))
VECTORS = os.path.abspath(os.path.join(HERE, "..", "..", "conformance", "mapping-vectors.json"))


def _segments(pointer):
    return [part.replace("~1", "/").replace("~0", "~") for part in pointer[1:].split("/")]


def _mutate(root, operation):
    parts = _segments(operation["path"])
    parent = root
    for part in parts[:-1]:
        parent = parent[int(part)] if isinstance(parent, list) else parent[part]
    key = int(parts[-1]) if isinstance(parent, list) else parts[-1]
    if operation["op"] == "delete":
        if isinstance(parent, list):
            parent.pop(key)
        else:
            del parent[key]
    elif operation["op"] == "set":
        parent[key] = copy.deepcopy(operation["value"])
    else:
        raise ValueError("unsupported vector mutation: " + operation["op"])


def _build_side(corpus, descriptor):
    profile = copy.deepcopy(corpus["profiles"][descriptor["profile"]])
    return {
        "source": copy.deepcopy(corpus["sources"][descriptor["source"]]),
        "profile": profile,
        "source_descriptor": copy.deepcopy(profile["source_format"]),
        "expected_profile_hash": (
            mapping_profile_hash(profile) if descriptor["pin"] == "profile" else descriptor["pin"]
        ),
        "native_verified": descriptor.get("native_verified", True),
    }


def run_mapping_vectors(corpus):
    results = []
    for vector in corpus["vectors"]:
        left = _build_side(corpus, vector["left"])
        right = _build_side(corpus, vector["right"])
        for operation in vector.get("mutations", []):
            side = left if operation["side"] == "left" else right
            _mutate(side[operation["target"]], operation)
        for side_name in vector.get("repin_after_mutation", []):
            side = left if side_name == "left" else right
            side["expected_profile_hash"] = mapping_profile_hash(side["profile"])
        result = compare_mapped_actions(
            left, right, definitions=corpus["definitions"], suite=corpus["suite"]
        )
        expected = vector["expect"]
        verdict_ok = result["verdict"] == expected["verdict"]
        if "reason_contains" in expected:
            reasons_ok = expected["reason_contains"] in result["reasons"]
        else:
            reasons_ok = result["reasons"] == expected.get("reasons", [])
        results.append(
            {
                "id": vector["id"],
                "pass": verdict_ok and reasons_ok,
                "verdict": result["verdict"],
                "reasons": result["reasons"],
            }
        )
    return results


with open(VECTORS, "r", encoding="utf-8") as handle:
    results = run_mapping_vectors(json.load(handle))

if "--json" in sys.argv:
    print(json.dumps(results, separators=(",", ":")))
else:
    for result in results:
        print(("PASS" if result["pass"] else "FAIL") + " " + result["id"] + " " + result["verdict"])

if any(not result["pass"] for result in results):
    raise SystemExit(1)

File: caid/impl/js/run-mapping-vectors.mjs
#!/usr/bin/env node
// SPDX-License-Identifier: Apache-2.0

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { compareMappedActions, mappingProfileHash } from './mapping.mjs';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const VECTORS = path.resolve(HERE, '../../conformance/mapping-vectors.json');
const clone = (value) => structuredClone(value);

function segments(pointer) {
  return pointer.slice(1).split('/').map((part) => part.replaceAll('~1', '/').replaceAll('~0', '~'));
}

function mutate(root, operation) {
  const parts = segments(operation.path);
  let parent = root;
  for (const part of parts.slice(0, -1)) parent = parent[Array.isArray(parent) ? Number(part) : part];
  const key = Array.isArray(parent) ? Number(parts.at(-1)) : parts.at(-1);
  if (operation.op === 'delete') {
    if (Array.isArray(parent)) parent.splice(key, 1);
    else delete parent[key];
  } else if (operation.op === 'set') {
    parent[key] = clone(operation.value);
  } else {
    throw new Error('unsupported vector mutation: ' + operation.op);
  }
}

function buildSide(corpus, descriptor) {
  const profile = clone(corpus.profiles[descriptor.profile]);
  const side = {
    source: clone(corpus.sources[descriptor.source]),
    profile,
    source_descriptor: clone(profile.source_format),
    expected_profile_hash: descriptor.pin === 'profile' ? mappingProfileHash(profile) : descriptor.pin,
    native_verified: descriptor.native_verified !== false,
  };
  return side;
}

export function runMappingVectors(corpus = JSON.parse(fs.readFileSync(VECTORS, 'utf8'))) {
  const results = [];
  for (const vector of corpus.vectors) {
    const left = buildSide(corpus, vector.left);
    const right = buildSide(corpus, vector.right);
    for (const operation of vector.mutations || []) mutate(operation.side === 'left' ? left[operation.target] : right[operation.target], operation);
    for (const sideName of vector.repin_after_mutation || []) {
      const side = sideName === 'left' ? left : right;
      side.expected_profile_hash = mappingProfileHash(side.profile);
    }
    const result = compareMappedActions(left, right, { definitions: corpus.definitions, suite: corpus.suite });
    const verdictOK = result.verdict === vector.expect.verdict;
    const reasonsOK = vector.expect.reason_contains
      ? result.reasons.includes(vector.expect.reason_contains)
      : JSON.stringify(result.reasons) === JSON.stringify(vector.expect.reasons || []);
    results.push({ id: vector.id, pass: verdictOK && reasonsOK, verdict: result.verdict, reasons: result.reasons });
  }
  return results;
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const results = runMappingVectors();
  if (process.argv.includes('--json')) {
    process.stdout.write(JSON.stringify(results) + '\n');
  } else {
    for (const result of results) console.log((result.pass ? 'PASS' : 'FAIL') + ' ' + result.id + ' ' + result.verdict);
  }
  if (results.some((result) => !result.pass)) process.exitCode = 1;
}
